### ISSUE TITLE
Unexpected Rugged::ZlibError in Commit#header_field

### ISSUE THREAD

**Issue by mastahyeti** [libgit2/rugged#564]
`rugged_exception_raise` seems to raise the wrong exception after a specific series of method calls.

1. `Repository#lookup`
1. `Repository#read_header`
1. `Commit#header_field` (with args that will cause an exception)

The final call should raise a `RuntimeError` afaict, but it actually raises a `Rugged::ZlibError`.

[Here's](https://github.com/libgit2/rugged/compare/master...mastahyeti:header_field_zlib_errors) some tests that demonstrate this behavior.

#563 actually fixes this specific case, because it catches the `GIT_ENOTFOUND` without calling `rugged_exception_check `, but it seems like there might be an underlying problem as well that should be fixed.

**Comment by arthurschreiber** [comment]
This is most likely coming from libgit2 internals. It looks like one of these calls is handling an error case but not resetting the error state in libgit2. So the next time `rugged_exception_check` is called, this stale error state is handled.

**Comment by carlosmn** [comment]
Indeed, in [commit.c:610](https://github.com/libgit2/libgit2/blob/master/src/commit.c#L610) we return an error code, but we do not set an error message, so rugged is grabbing whatever existed just before. The "real" fix would be to add a `giterr_set()` call here about not finding the header field.

**Comment by mastahyeti** [comment]
Looks like it's a pattern in libgit2 to return `GIT_ENOTFOUND` without calling `giterr_set()`. I guess https://github.com/libgit2/rugged/pull/563 might actually be the correct fix for this bug.
